### PR TITLE
Sprint 1 - User Sevice Created

### PR DESCRIPTION
**Pull request by Akash-Ping-NT**

### CODE REVIEW

File: Restaurant_Service/src/main/java/com/microservice/RestaurantService/Service/ServiceImpl/RestaurantServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+import java.util.Optional;
+
+@Service
+public class RestaurantServiceImpl implements RestaurantService {
+
+    private final RestaurantRepository restaurantRepository;
+    private final UserServiceClient userServiceClient;
+
+    @Autowired
+    public RestaurantServiceImpl(RestaurantRepository restaurantRepository, UserServiceClient userServiceClient) {
```
Comment (alok-nucleusteq) [inline Restaurant_Service/src/main/java/com/microservice/RestaurantService/Service/ServiceImpl/RestaurantServiceImpl.java]: write complete impl - implementation

File: Restaurant_Service/src/main/java/com/microservice/RestaurantService/Service/ServiceImpl/RestaurantServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    @Override
+    public CreateRestaurantResponse createRestaurant(String email, CreateRestaurantRequest createRestaurantRequest) {
+
+        UserDtoResponse userDtoResponse = userServiceClient.getUserByEmail(email);
+
+        // Check if the user role is allowed to create a restaurant (e.g., OWNER role)
+        if (!"RESTAURANT_OWNER".equals(userDtoResponse.getRole())) {
+            throw new IllegalArgumentException("User is not authorized to create a restaurant.");
```
Comment (alok-nucleusteq) [inline Restaurant_Service/src/main/java/com/microservice/RestaurantService/Service/ServiceImpl/RestaurantServiceImpl.java]: IllegalArgumentException use custom exception that is related to the error like unauthorized exception

File: Restaurant_Service/src/main/java/com/microservice/RestaurantService/Service/ServiceImpl/RestaurantServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        // Create a new Restaurant entity
+        Restaurant restaurant = new Restaurant();
+        restaurant.setAddress(createRestaurantRequest.getAddress());
+        restaurant.setEmail(createRestaurantRequest.getEmail());
+        restaurant.setRestaurantName(createRestaurantRequest.getRestaurantName());
+        restaurant.setContactNo(createRestaurantRequest.getContactInformation());
+        restaurant.setOpen(createRestaurantRequest.getOpen());
+        restaurant.setOpeningHours(createRestaurantRequest.getOpeningHours());
+        restaurant.setRegistrationDate(createRestaurantRequest.getRegistrationDate());
+        restaurant.setOwnerId(userDtoResponse.getId());
```
Comment (alok-nucleusteq) [inline Restaurant_Service/src/main/java/com/microservice/RestaurantService/Service/ServiceImpl/RestaurantServiceImpl.java]: Remove unnecessary comments

File: Restaurant_Service/src/main/java/com/microservice/RestaurantService/controller/RestaurantController.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+import com.microservice.RestaurantService.dtos.request.CreateRestaurantRequest;
+import com.microservice.RestaurantService.dtos.response.CreateRestaurantResponse;
+import com.microservice.RestaurantService.exceptions.RestaurantCreationException;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+import org.springframework.web.server.ResponseStatusException;
+
+@RestController
```
Comment (alok-nucleusteq) [inline Restaurant_Service/src/main/java/com/microservice/RestaurantService/controller/RestaurantController.java:13]: add java docs

File: Restaurant_Service/src/main/java/com/microservice/RestaurantService/dtos/request/CreateRestaurantRequest.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+public class CreateRestaurantRequest {
+
+    private String restaurantName;
+    private String email;
+    private String address;
+    private String contactInformation;
+    private Boolean open;
+    private String openingHours;
+    private LocalDateTime registrationDate;
```
Comment (alok-nucleusteq) [inline Restaurant_Service/src/main/java/com/microservice/RestaurantService/dtos/request/CreateRestaurantRequest.java]: Remove unnecessary comments

File: Restaurant_Service/src/main/java/com/microservice/RestaurantService/dtos/request/CreateRestaurantRequest.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    public void setEmail(String email) {
+        this.email = email;
+    }
+
+    public String getRestaurantName() {
+        return restaurantName;
+    }
+
```
Comment (alok-nucleusteq) [inline Restaurant_Service/src/main/java/com/microservice/RestaurantService/dtos/request/CreateRestaurantRequest.java]: Implement hashcode and equal method in every entity and dtos

File: Restaurant_Service/src/main/java/com/microservice/RestaurantService/dtos/request/CreateRestaurantRequest.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    public void setRestaurantName(String restaurantName) {
+        this.restaurantName = restaurantName;
+    }
```
Comment (alok-nucleusteq) [inline Restaurant_Service/src/main/java/com/microservice/RestaurantService/dtos/request/CreateRestaurantRequest.java]: Implement toString method in dtos only

File: Restaurant_Service/src/main/java/com/microservice/RestaurantService/dtos/response/CreateRestaurantResponse.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+
+    public String getRestaurantName() {
+        return restaurantName;
+    }
+
+    public void setRestaurantName(String restaurantName) {
+        this.restaurantName = restaurantName;
+    }
+
+    public String getOwnerName() {
```
Comment (alok-nucleusteq) [inline Restaurant_Service/src/main/java/com/microservice/RestaurantService/dtos/response/CreateRestaurantResponse.java]: In entity implement hascode and equal method only not toString

File: Restaurant_Service/src/main/java/com/microservice/RestaurantService/entity/Restaurant.java [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+public class Restaurant {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @Column(nullable = false)
+    private String restaurantName;
+
+    @Column(nullable = false, unique = true)
```
Comment (alok-nucleusteq) [inline Restaurant_Service/src/main/java/com/microservice/RestaurantService/entity/Restaurant.java:18]: validate email with regular expression (format should be only akash.ping@nucleusteq.com)

File: Restaurant_Service/src/main/java/com/microservice/RestaurantService/entity/Restaurant.java [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+import java.util.Objects;
+
+@Entity
+@Table(name = "restaurant")
+public class Restaurant {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
```
Comment (alok-nucleusteq) [inline Restaurant_Service/src/main/java/com/microservice/RestaurantService/entity/Restaurant.java:14]: name should be only in string numbers not allowed

File: Restaurant_Service/src/main/java/com/microservice/RestaurantService/entity/Restaurant.java [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+    @Column(nullable = false)
+    private String restaurantName;
+
+    @Column(nullable = false, unique = true)
+    private String email;
+
+    @Column(nullable = false)
+    private String address;
+
+    @Column(name = "contact_no", nullable = false)
```
Comment (alok-nucleusteq) [inline Restaurant_Service/src/main/java/com/microservice/RestaurantService/entity/Restaurant.java:24]: Only number should be allowed and should not be more than 10 digits

File: Restaurant_Service/src/main/java/com/microservice/RestaurantService/entity/Restaurant.java [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+                Objects.equals(openingHours, that.openingHours) &&
+                Objects.equals(registrationDate, that.registrationDate) &&
+                Objects.equals(ownerId, that.ownerId);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id, restaurantName, email, address, contactNo, open, openingHours, registrationDate, ownerId);
+    }
+
```
Comment (alok-nucleusteq) [inline Restaurant_Service/src/main/java/com/microservice/RestaurantService/entity/Restaurant.java:148]: toString method should be in dtos only

File: Restaurant_Service/src/main/java/com/microservice/RestaurantService/package-info.java [hunk 13]
```diff
@@ -0,0 +1,4 @@
+/**
```
Comment (alok-nucleusteq) [inline Restaurant_Service/src/main/java/com/microservice/RestaurantService/package-info.java:1]: remove this file from here add checkstyle rule to ignore this error

File: Restaurant_Service/src/test/java/com/microservice/RestaurantService/RestaurantServiceApplicationTests.java [hunk 14]
```diff
@@ -0,0 +1,10 @@
+package com.microservice.RestaurantService;
+
+import org.springframework.boot.test.context.SpringBootTest;
+
+@SpringBootTest
```
Comment (alok-nucleusteq) [inline Restaurant_Service/src/test/java/com/microservice/RestaurantService/RestaurantServiceApplicationTests.java:5]: implement properly (for service use mockito)

File: userService/src/main/java/com/microservice/userService/controller/UserController.java [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+     * @return ResponseEntity indicating status of the authentication operation.
+     */
+    @PostMapping("/login")
+    public ResponseEntity<UserLoginResponseDTO> loginUser(@Valid @RequestBody final UserLoginRequestDTO loginRequest) {
+        LOGGER.info("Starting the login Functionality");
+        try {
+            UserLoginResponseDTO loginResDto = userService.loginUser(loginRequest);
+            return new ResponseEntity<>(loginResDto, HttpStatus.OK);
+        } catch (Exception e) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Login Failed Check Credential", e);
```
Comment (alok-nucleusteq) [inline userService/src/main/java/com/microservice/userService/controller/UserController.java]: implement logger in the end of the method also before ending the execution

File: userService/src/main/java/com/microservice/userService/controller/UserController.java [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+        }
+    }
+
+    /**
+     * Handles the address addition of a user .
+     * @param addressAddDTO The DTO containing the login credentials.
+     * @return ResponseEntity indicating status of the authentication operation.
+     */
+    @PostMapping("/{userId}/addAddresses")
+    public ResponseEntity<UserAddressResponseDTO> addAddress(
```
Comment (alok-nucleusteq) [inline userService/src/main/java/com/microservice/userService/controller/UserController.java]: logger should be implemented in every method of controller and service 
(in the start and end of the method)

File: userService/src/main/java/com/microservice/userService/controller/UserController.java [hunk 17]
```diff
@@ -0,0 +1,138 @@
...
+            @PathVariable final Long userId,
+            @RequestBody final UserAddressRequestDTO addressAddDTO) {
+        try {
+            UserAddressResponseDTO responseDTO = deliveryAddressService.addAddress(userId, addressAddDTO);
+            return new ResponseEntity<>(responseDTO, HttpStatus.CREATED);
+        } catch (Exception e) {
+            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Failed to add address", e);
```
Comment (alok-nucleusteq) [inline userService/src/main/java/com/microservice/userService/controller/UserController.java]: Use proper exception related to issues

File: userService/src/main/java/com/microservice/userService/controller/UserController.java [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+    }
+
+    @PutMapping("/{addressId}/update")
+    public ResponseEntity<CommonResponseDTO> updateAddress(
+            @PathVariable final Long addressId,
+            @RequestBody final UserAddressRequestDTO addressUpdateDTO) {
+        try {
+            CommonResponseDTO responseDTO = deliveryAddressService.updateAddress(addressId, addressUpdateDTO);
+            return ResponseEntity.ok(responseDTO);
+        } catch (UserNotFoundException e) {
```
Comment (alok-nucleusteq) [inline userService/src/main/java/com/microservice/userService/controller/UserController.java:111]: Here should be only one exception multiple exception you can throw in service

File: userService/src/main/java/com/microservice/userService/controller/package-info.java [hunk 19]
```diff
@@ -0,0 +1,4 @@
+/**
```
Comment (alok-nucleusteq) [inline userService/src/main/java/com/microservice/userService/controller/package-info.java]: remove package info files

File: userService/src/main/java/com/microservice/userService/dtos/request/UserAddressRequestDTO.java [hunk 20]
```diff
@@ -0,0 +1,78 @@
+package com.microservice.userService.dtos.request;
+
+public class UserAddressRequestDTO {
+
+    private Long id;
```
Comment (alok-nucleusteq) [inline userService/src/main/java/com/microservice/userService/dtos/request/UserAddressRequestDTO.java]: Add java docs and validation

File: userService/src/main/java/com/microservice/userService/dtos/request/UserLoginRequestDTO.java [hunk 21]
```diff
@@ -0,0 +1,30 @@
+package com.microservice.userService.dtos.request;
+
+import javax.validation.constraints.Email;
+import javax.validation.constraints.NotBlank;
+
+public class UserLoginRequestDTO {
+
+    @NotBlank(message = "Email is mandatory")
```
Comment (alok-nucleusteq) [inline userService/src/main/java/com/microservice/userService/dtos/request/UserLoginRequestDTO.java:18]: validate email

File: userService/src/main/java/com/microservice/userService/dtos/request/UserRegistrationRequestDTO.java [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+import javax.validation.constraints.Email;
+import javax.validation.constraints.NotBlank;
+
+public class UserRegistrationRequestDTO {
+
+    @NotBlank(message = "Name is mandatory")
+    private String name;
+
+    @NotBlank(message = "Email is mandatory")
+    @Email(message = "Email should be valid")
```
Comment (alok-nucleusteq) [inline userService/src/main/java/com/microservice/userService/dtos/request/UserRegistrationRequestDTO.java:32]: where are you checking image is valid ? ("Email should be valid")

File: userService/src/main/java/com/microservice/userService/dtos/response/CommonResponseDTO.java [hunk 23]
```diff
@@ -0,0 +1,18 @@
+package com.microservice.userService.dtos.response;
```
Comment (alok-nucleusteq) [inline userService/src/main/java/com/microservice/userService/dtos/response/CommonResponseDTO.java:1]: Add java docs

File: userService/src/main/java/com/microservice/userService/dtos/response/UserRegistrationResponseDTO.java [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+
+import java.math.BigDecimal;
+
+public class UserRegistrationResponseDTO {
+
+    private Long userId;
+    private String name;
+    private String email;
+    private String role;
+    private String message; // A message to indicate successful registration
```
Comment (alok-nucleusteq) [inline userService/src/main/java/com/microservice/userService/dtos/response/UserRegistrationResponseDTO.java]: remove comments add in javadoc

File: userService/src/main/java/com/microservice/userService/entity/User.java [hunk 25]
```diff
@@ -0,0 +1,136 @@
...
+    private String name;
+
+    @Column(nullable = false, unique = true, length = 255)
+    private String email;
+
+    @Enumerated(EnumType.STRING)
+    @Column(nullable = false)
+    private Role role;
+
+    @Column(nullable = false, length = 255)
```
Comment (alok-nucleusteq) [inline userService/src/main/java/com/microservice/userService/entity/User.java]: follow above reviews

**Review by alok-nucleusteq — CHANGES_REQUESTED** [review]
Resolve the reviews asap

File: Restaurant_Service/src/main/java/com/microservice/RestaurantService/Service/ServiceImpl/RestaurantServiceImpl.java (hunk 1 above)
Comment (Akash-Ping-NT) [reply Restaurant_Service/src/main/java/com/microservice/RestaurantService/Service/ServiceImpl/RestaurantServiceImpl.java]: Yes Alok , I have made the changes.

File: Restaurant_Service/src/main/java/com/microservice/RestaurantService/Service/ServiceImpl/RestaurantServiceImpl.java (hunk 2 above)
Comment (Akash-Ping-NT) [reply Restaurant_Service/src/main/java/com/microservice/RestaurantService/Service/ServiceImpl/RestaurantServiceImpl.java]: Done with implementing custom exception.

File: Restaurant_Service/src/main/java/com/microservice/RestaurantService/Service/ServiceImpl/RestaurantServiceImpl.java (hunk 3 above)
Comment (Akash-Ping-NT) [reply Restaurant_Service/src/main/java/com/microservice/RestaurantService/Service/ServiceImpl/RestaurantServiceImpl.java]: Implemented

File: Restaurant_Service/src/main/java/com/microservice/RestaurantService/controller/RestaurantController.java (hunk 4 above)
Comment (Akash-Ping-NT) [reply Restaurant_Service/src/main/java/com/microservice/RestaurantService/controller/RestaurantController.java:13]: Done with adding java docs.

**Pull request merged**